training/controllers/api.py

```diff
@@ -0,0 +1,200 @@
+import json
+import os
+from flask import Blueprint, jsonify, abort, make_response, request, session
+from itsdangerous import URLSafeTimedSerializer
+
+from training.controllers.function_decorators import login_required
+from training.utils.common_functions import how_is_logged
+from training.extensions import bcrypt, db
+from training.models.users import User
+
+bp = Blueprint('api', __name__, url_prefix='/api')
+
+serializer = URLSafeTimedSerializer(os.environ.get('SECRET_KEY'))
+
+tasks = [
+    {
+        'id': 1,
+        'title': u'Buy groceries',
+        'description': u'Milk, Cheese, Pizza, Fruit, Tylenol',
+        'done': False
+    },
+    {
+        'id': 2,
+        'title': u'Learn Python',
+        'description': u'Need to find a good Python tutorial on the web',
+        'done': False
+    },
+    {
+        'id': 3,
+        'title': u'Learn Python',
+        'description': u'Need to find a good Python tutorial on the web',
+        'done': True
+    }
+]
+
+
+@bp.route('/example')
+def example():
+    return jsonify({'tasks': tasks})
+
+
+@bp.route('/example/<int:task_id>', methods=['GET'])
+def get_task(task_id):
+    task = [task for task in tasks if task['id'] == task_id]
+    if len(task) == 0:
+        # 404 means resource not found, exactly what happens here
+        abort(404)
+    return jsonify({'task': task[0]})
+
+
+"""
+To avoid this response when a 404 error was generated:
+<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
+<title>404 Not Found</title>
+<h1>Not Found</h1>
+<p>The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.</p>
+ 
+ We craft the function below.
+"""
+
+
+@bp.errorhandler(404)
+def not_found(error):
+    return make_response(jsonify({'error': 'Not found doggy'}), 404)
+
+
+@bp.route('/example', methods=['POST'])
+def create_task():
+    if not request.json or not 'title' in request.json:
+        print("Mamarracho")
+        abort(400)
+    else:
+        task = {
+            'id': tasks[-1]['id'] + 1,
+            'title': request.json['title'],
+            'description': request.json.get('description', ""),
+            'done': False
+        }
+        tasks.append(task)
+        return jsonify({'task': task}), 201
+
+    session['title'] = request.json['title']
+    return "PEPE", 205
+
+# request.get_json()
+
+
+# True endpoint start here
+def repeated_fields(fields):
+    if User.query.filter_by(id=fields.get('username')).first() is not None or User.query.filter_by(id=fields.get('email')).first() is not None:
+        return True
+    return False
+
+
+@bp.route('/register', methods=['POST'])
+def sign_up():
+    if not request.json or not 'username' in request.json or not 'password' in request.json or not 'email' in request.json:
+        abort(400)
```

Comment:
I'd be nice to let the API client know which are the errors. This applies to every 4XX status code (because are meant for client errors)

Reply:
I changed every return statement to JSON Response and checked all status codes